api/envoy/config/route/v3/route_components.proto

```diff
@@ -320,9 +320,35 @@ message WeightedCluster {
 
     reserved "per_filter_config";
 
-    // Name of the upstream cluster. The cluster must exist in the
-    // :ref:`cluster manager configuration <config_cluster_manager>`.
-    string name = 1 [(validate.rules).string = {min_len: 1}];
+    // Only one of *name* and *cluster_header* may be specified.
+    oneof cluster_specifier {
+      option (validate.required) = true;
+
+      // Name of the upstream cluster. The cluster must exist in the
+      // :ref:`cluster manager configuration <config_cluster_manager>`.
+      string name = 1 [
```

Comment:
You can't move this existing field into a oneof, because that's a breaking API change.   You need to leave this outside of the oneof for now.

The purpose of the `(udpa.annotations.field_migrate).oneof_promotion` annotation is to tell us to move these fields into the oneof if/when we ever release a new version of this API.  That's basically what you're supposed to use in this kind of case instead of adding a oneof.

So basically, I would just do the following here:

1. Add the `(udpa.annotations.field_migrate).oneof_promotion` annotation to the existing `name` field.
2. Add the new `cluster_header` field with the same annotation.
3. Add comments for both fields saying that only one of the two may be set.

Reply:
Thanks for the review! I have thought about backwards compatibility but looks like Envoy's rule is stricter even though it is safe in terms of protobuf wire compatibility

If we leave this outside of `oneof`,  I think we will need to remove validate.rules on `name` and `cluster_header` for now as well. Otherwise, Proto constraint validation will fail.

I updated the code as described above but I am new to Envoy API, if you have any other good ideas, please let me know.

Reply:
Moving a field into a oneof is fine for wire compatibility, but it will cause API changes in the generated code in golang.  Since we promise both wire compatibility and stability for build dependency for API clients other than Envoy (such as grpc-go), that would be a breaking change.

I suspect that removing the proto-gen-validate constraint is okay, as long as we add a check to the implementation to ensure that one of these two fields is set.  @envoyproxy/api-shepherds, can someone confirm that that's correct?

Reply:
@htuch says that Envoy will need to do explicit validation of these fields, so he'd prefer if the Envoy implementation was made part of this PR.  Otherwise, we will lose the PGV check without yet having the new explicit check.

Reply:
Make sense.
In the spirit of "small cl", I have added the minimal implementation + validation logic and its unit tests (rather than adding full CL e.g. with weighted cluster pick logic, integration test, etc).

Please let me what do you think.